library/alloc/src/boxed.rs

```diff
@@ -518,20 +751,47 @@ impl<T: ?Sized> Box<T> {
     /// ```
     #[stable(feature = "box_leak", since = "1.26.0")]
     #[inline]
-    pub fn leak<'a>(b: Box<T>) -> &'a mut T
+    pub fn leak<'a>(b: Self) -> &'a mut T
     where
-        T: 'a, // Technically not needed, but kept to be explicit.
+        A: 'a,
     {
         unsafe { &mut *mem::ManuallyDrop::new(b).0.as_ptr() }
     }
 
+    /// Consumes and leaks the `Box`, returning a mutable reference,
+    /// `&'a mut T`, and the allocator. Note that the type `T` must outlive the chosen lifetime
+    /// `'a`. If the type has only static references, or none at all, then this
+    /// may be chosen to be `'static`.
+    ///
+    /// This function is mainly useful for data that lives for the remainder of
+    /// the program's life. Dropping the returned reference will cause a memory
+    /// leak. If this is not acceptable, the reference should first be wrapped
+    /// with the [`Box::from_raw`] function producing a `Box`. This `Box` can
+    /// then be dropped which will properly destroy `T` and release the
+    /// allocated memory.
+    ///
+    /// Note: this is an associated function, which means that you have
+    /// to call it as `Box::leak(b)` instead of `b.leak()`. This
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn leak_with_alloc<'a>(b: Self) -> (&'a mut T, A)
```

Comment:
This API seems sketchy to me. Without thinking too much about it I think one could do something along the lines of:

```rust
let (ref, allocref) = Box::leak_with_alloc(a_box);
drop(allocref); // potentially drops the backing storage
use(ref); // use of a dangling reference
```

Reply:
That's indeed a problem. Do you have a solution to this or should I remove this method? Therefore also a comment should be added to `into_raw_with_alloc`, that the allocator may have to outlive the pointer.

Reply:
I agree with @nagisa, for `leak` to be sound you need to actually leak the allocator, not return it to the user. I'd just remove that method.

`into_raw_with_alloc` is fine since it returns a raw pointer and it's not our problem if the user misuses it.